src/components/ProfilePage.jsx

```diff
@@ -9,152 +9,78 @@ import {
+import Certificates from './ProfilePage/Certificates';
 
 export class ProfilePage extends React.Component {
   constructor(props) {
     super(props);
 
-    this.state = {
-      fullName: { value: null, visibility: null },
-      userLocation: { value: null, visibility: null },
-      education: { value: null, visibility: null },
-      bio: { value: null, visibility: null },
-      socialLinks: { value: null, visibility: null },
-      certificates: { value: null, visibility: null },
-    };
-
-    this.onCancel = this.onCancel.bind(this);
-    this.onEdit = this.onEdit.bind(this);
-    this.onSave = this.onSave.bind(this);
-    this.onSaveProfilePhoto = this.onSaveProfilePhoto.bind(this);
-    this.onDeleteProfilePhoto = this.onDeleteProfilePhoto.bind(this);
-    this.onChange = this.onChange.bind(this);
-    this.onVisibilityChange = this.onVisibilityChange.bind(this);
+    this.handleSaveProfilePhoto = this.handleSaveProfilePhoto.bind(this);
+    this.handleDeleteProfilePhoto = this.handleDeleteProfilePhoto.bind(this);
+    this.handleClose = this.handleClose.bind(this);
+    this.handleOpen = this.handleOpen.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleChange = this.handleChange.bind(this);
   }
 
   componentDidMount() {
     this.props.fetchProfile(this.props.match.params.username);
   }
 
-  onCancel() {
-    this.props.closeField(this.props.currentlyEditingField);
+  handleSaveProfilePhoto(formData) {
+    this.props.saveProfilePhoto(this.props.username, formData);
   }
 
-  onEdit(fieldName) {
-    this.props.openField(fieldName);
+  handleDeleteProfilePhoto() {
+    this.props.deleteProfilePhoto(this.props.username);
   }
 
-  onSave(fieldName) {
-    const {
-      value,
-      visibility,
-    } = this.state[fieldName];
-
-    const data = {};
-
-    if (value != null) {
-      data.profileData = { [fieldName]: value };
-    }
-    if (visibility != null) {
-      data.preferencesData = { visibility: { [fieldName]: visibility } };
-    }
-
-    if (value == null && visibility == null) {
-      this.onCancel();
-    } else {
-      this.props.saveProfile(this.props.username, data, fieldName);
-    }
+  handleClose(formId) {
+    this.props.closeForm(formId);
   }
 
-  onSaveProfilePhoto(formData) {
-    this.props.saveProfilePhoto(this.props.username, formData);
+  handleOpen(formId) {
+    this.props.openForm(formId);
   }
 
-  onDeleteProfilePhoto() {
-    this.props.deleteProfilePhoto(this.props.username);
+  handleSubmit(formId) {
+    this.props.saveProfile(formId);
   }
 
-  onChange(fieldName, value) {
-    this.setState({
-      [fieldName]: {
-        value,
-        visibility: this.state[fieldName].visibility,
-      },
-    });
-  }
-
-  onVisibilityChange(fieldName, visibility) {
-    this.setState({
-      [fieldName]: {
-        value: this.state[fieldName].value,
-        visibility,
-      },
-    });
+  handleChange(formId, name, value) {
+    if (name === 'visibility') {
```

Comment:
Depending on what kind of field changed, we store the data in different places in the reducer.  Separate actions felt like a relatively clean way to tell the difference.